contents/javascript/Javascript-promises.md

```diff
@@ -90,57 +86,118 @@ These situations correspond to the 3 states of Promises in Javascript.
 
-After defining the function, we call it, and we call another function `.then()` upon the result. `.then()` is used to consume the result from a fulfilled Promise and act on it. In this case, we are using it to log our result. More importantly, it lets us chain promises to avoid the deep nesting known as callback hell.
+### The iterative style of promises
 
-Here's the callback hell code rewritten with Promises.
+Using promises is the functional programming way of writing asynchronous code. Data is passed from one `.then()` call to the next. However, this style of programming requires some familiarity in modelling programming instructions as changes in the data. 
 
-```javascript
+That is unlike the iterative programming style which is considered to be a more intuitive way of writing code. As such, the `async` and `await` keywords are implemented to facilitate a iterative way of using promises.
 
-function update(numTimes) {
-    return new Promise((resolve, reject) => {
-        var x = "I have not been updated";
-        setTimeout(function() {
-            x = "I have been updated " + numTimes + " times.";
-            console.log(x);
-            resolve(numTimes);
-        }, 1000);
-    });
-}
+The `async` keyword ensures that the function returns a promise, and as for `await`, you can think of it as waiting for some promise to return the value before continuing. So back to the same example.
 
-update(1).then((result) => update(result + 1))
-.then((result) => update(result + 1));
+So here is the code in the "promise style", as a recap.
 
+```javascript
+/** func implementation same as above **/
+// getData()
+// intenseWait(val)
+// filterData(data, callback)
+
+/** changes only required in main **/
+function main() {
+    return getData()
+        .then(data => filterData(data))
+        .then(fltered => console.log(filtered));
+}
+
+main();
 ```
 
-Now we update our function to take in a parameter `numTimes` so that we can log how many times `x` was updated. `then()` will return a Promise, which you can then again call `.then()`. By doing this, we can chain our Promises and ensure that execution of the code goes in the order that we require. In addition, it also allows your variables to be correctly passed down, ensuring that you will not have to deal with any undefined variables along the way.
+Using the "async await" style, only the `main()` function needs to be changed. After the change, you will realize that the code becomes much simpler to understand. There is a clear improvement in clarity and readability.
 
-Using Promises helps your code to be clean and easy to read, which makes it easy for anyone to instantly know what the code does, instead of having to trace through the code written with callbacks.
+```javascript
+async function main() {
+    const data = await getData();
+    const filtered = await filterData(data);
 
-## How Promises are used
+    console.log(filtered);
+}
+```
```

Comment:
Can be simplified as follows:
> It can we rewritten in the 'async-await' style as follows.
```javascript
async function main() {
    const data = await getData();
    const filtered = await filterData(data);
    console.log(filtered);
}
```